### DISCUSSION TITLE
9 - Note Lifecycle Plugins

### DISCUSSION THREAD

**Discussion by kevinslin** [dendronhq/dendron#680] [RFCs/Ideas]
This thread is to discuss the [note lifecycles rfc](https://wiki.dendron.so/notes/d2f8fe67-36c7-4600-b745-c22bdcb5b2cf.html)

**Comment by codefriar** [comment]
In general I love this. 

The RFC discusses registering plugins, and hints at lifecycle hooks those plugins could use.

I'd like to suggest that plugins and lifecycle hooks should be two separate things. Or, at least, that the lifecycle hooks and plugins can be 'attached' to schemas. Ie: I want my daily.journal.* notes to execute plugin.foo on create. However, I don't want plugin.foo to run when I create a note of schema 'meeting'. 

This would, I think, reduce the latency concern expressed in the RFC, as the fine-grained detail of what runs when will prevent a ton of plugins from all trying to execute every note creation.

It might be helpful to enumerate the lifecycle events in the RFC:

* OnCreate
* OnSave
* OnDelete
* OnCreateWithSchema
* OnSaveWithSchema
* OnDeleteWithSchmea...

**Reply by kevinslin** [reply]
Thanks for the input. My takeaways:
- nomencalture for lifecycle, hooks and plugins
    - realize this is a bit confusing
        - lifecycle describes the flow of events that occur when dendron performs operations on notes
        - hooks are custom logic that can be added at specific points of the lifecycle
    - maybe we can just do away with `plugins` and call it hooks 🤔
- integration of hooks with schemas
    - your right, this wasn't mentioned at all in the rfc
    - we would extend the schema syntax to support hooks
        ```yml
        - id: foo
        hooks:
            onCreate:
            - id: addEmoji
        ```
    - our current use case of applying a template can then be refactored as a `hook` that runs during `onCreate`
- we should be able to differentiate when we want to run a hook on a schema vs a note. we could add another property when registering a hook
    ```yml
    type: schema|note|all
    ```
- enumerating of events:
    - will add the enumeration

**Comment by brettmitchelldev** [comment]
Coming over from https://github.com/dendronhq/dendron/issues/2270. I figured since there's an open discussion on this topic I'd describe my specific use case for hooks. Hopefully someone finds it helpful (or at least mildly interesting). :shrug:

I'm definitely not a Dendron expert yet, especially WRG the plans for it's future, so maybe there's a different feature coming down the pipe (or already here) that would be a better fit for this, but one use case I'd like to use hooks for is dynamic views.

Maybe it's a bit hacky, but personally, I'd like to be able to dynamically populate a note every time I open it, or better yet, define a set of ephemeral notes that don't exist unless they are open (if this never becomes a first-class citizen, I'd probably use an onOpen / onClose pair to populate and then delete).

Reason being is I have some infrastructure set up to manage JS functions in code blocks within notes that can be designated as 'contributers' for a particular hierarchy pattern. Right now, this is handy for contributing todo items to a daily journal note from locations within my 'projects.*' hierarchy in ways that a schema with templates can't handle. I'd like to extend this to provide myself with dynamic views. I'm not quite sure yet how the relationship between an ephemeral dynamic view and the note hierarchy would work :shrug: but that's the general direction I'm headed in.

All that to say, I'm very excited for an expanded hooks API, and if anyone has pointers (or criticisms) of my process, I'm happy to hear them.
